src/Shell.zig

```diff
@@ -0,0 +1,303 @@
+        error.IsDir => return true,
+        else => return err,
+    };
+
+    return stat.kind == .Directory;
+}
+
+/// Analogue of the `find` utility, returns a set of paths matching filtering criteria.
+///
+/// Returned slice is stored in `Shell.arena`.
+pub fn find(shell: *Shell, options: struct {
+    where: []const []const u8,
+    ends_with: []const u8,
+}) ![]const []const u8 {
+    var result = std.ArrayList([]const u8).init(shell.arena.allocator());
+
+    const cwd = std.fs.cwd();
+    for (options.where) |base_path| {
+        var base_dir = try cwd.openDir(base_path, .{ .iterate = true });
+        defer base_dir.close();
+
+        var walker = try base_dir.walk(shell.gpa);
+        defer walker.deinit();
+
+        while (try walker.next()) |entry| {
+            if (std.mem.endsWith(u8, entry.path, options.ends_with)) {
+                const full_path =
+                    try std.fs.path.join(shell.arena.allocator(), &.{ base_path, entry.path });
+                try result.append(full_path);
+            }
+        }
+    }
+
+    return result.items;
+}
+
+/// Runs the given command with inherited stdout and stderr.
+/// Returns an error if exit status is non-zero.
+///
+/// Supports interpolation using the following syntax:
+///
+/// ```
+/// shell.exec("git branch {op} {branches}", .{
+///     op = "-D",
+///     branches = &.{"main", "feature"},
+/// })
+/// ```
+pub fn exec(shell: Shell, comptime cmd: []const u8, args: anytype) !void {
+    var argv = Argv.init(shell.gpa);
+    defer argv.deinit();
+
+    try format_argv(&argv, cmd, args);
+
+    const child = try std.ChildProcess.init(argv.items, shell.gpa);
+    defer child.deinit();
+
+    child.stdin_behavior = .Ignore;
+    child.stdout_behavior = .Inherit;
+    child.stderr_behavior = .Inherit;
+
+    const term = try child.spawnAndWait();
+
+    switch (term) {
+        .Exited => |code| if (code != 0) return error.NonZeroExitStatus,
+        else => return error.CommandFailed,
+    }
+}
+
+/// Returns `true` if the command executed successfully with a zero exit code and an empty stderr.
+///
+/// The intended use-case is sanity-checking that an executable is present, by running
+/// `my-tool --version`.
+pub fn exec_status_ok(shell: *Shell, comptime cmd: []const u8, args: anytype) !bool {
+    var argv = Argv.init(shell.gpa);
+    defer argv.deinit();
+
+    try format_argv(&argv, cmd, args);
+
+    const res = std.ChildProcess.exec(.{
+        .allocator = shell.gpa,
+        .argv = argv.items,
+    }) catch return false;
+    defer shell.gpa.free(res.stderr);
+    defer shell.gpa.free(res.stdout);
+
+    return switch (res.term) {
+        .Exited => |code| code == 0 and res.stderr.len == 0,
+        else => false,
+    };
+}
+
+/// Run the command and return its stdout.
+///
+/// Returns an error if the command exists with a non-zero status or a non-empty stderr.
+pub fn exec_stdout(shell: *Shell, comptime cmd: []const u8, args: anytype) ![]const u8 {
```

Comment:
Yeah, the slightly sprawling design here is intentional. I've added an explanatory comment to the start of the file. There's already a fully-flexible, give-caller-all-control API in std.ChildProcess. The problem with that is that the caller has too much control over the redirection, status codes, etc, and it's easy to forget something. 

So here we rather try to orient API towards specific use-cases, so that the caller doesn't have any knobs, and needs to just select the correct function.


Now, how exactly do we turn the knobs is an open question (should we check if stderr is zero? would it be a good idea to echo the command itself in `exec`?), but I think we can adjust that as we go.